pkg/ccl/importccl/import_stmt.go

```diff
@@ -712,6 +608,179 @@ type importResumer struct {
+	err := txn.CPut(ctx, sqlbase.MakeDescMetadataKey(desc.ID),
+		sqlbase.WrapDescriptor(&importing), sqlbase.WrapDescriptor(desc))
+	if err != nil {
+		return nil, errors.Wrap(err, "another operation is currently operating on the table")
+	}
+
+	return &importing, nil
+	// NB: we need to wait for the schema change to show up before it is safe
+	// to ingest, but rather than do that here, we'll wait for this schema
+	// change in the job's Resume hook, before running the ingest phase. That
+	// will hopefully let it get a head start on propagating, plus the more we
+	// do in the job, the more that has automatic cleanup on rollback.
+}
+
+// prepareTableDescsForIngestion prepares table descriptors for the ingestion
+// step of import. The descriptors are in an IMPORTING state (offline) on
+// successful completion of this method.
+func (r *importResumer) prepareTableDescsForIngestion(
+	ctx context.Context, p sql.PlanHookState, details jobspb.ImportDetails,
+) error {
+	importDetails := details
+	var hasExistingTables bool
+	err := p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
+		var err error
+		newTableDescToIdx := make(map[*sqlbase.TableDescriptor]int, len(importDetails.Tables))
+		var newTableDescs []jobspb.ImportDetails_Table
+		for i, table := range importDetails.Tables {
+			if !table.IsNew {
+				importDetails.Tables[i].Desc, err = prepareExistingTableDescForIngestion(ctx, txn, table.Desc, p)
+				if err != nil {
+					return err
+				}
+				hasExistingTables = true
+			} else {
+				newTableDescToIdx[table.Desc] = i
+				newTableDescs = append(newTableDescs, table)
+			}
+		}
+
+		// Prepare the table descriptors for newly created tables being imported
+		// into.
+		//
+		// TODO(adityamaru): This is still unnecessarily complicated. If we can get
+		// the new table desc preparation to work on a per desc basis, rather than
+		// requiring all the newly created descriptors, then this can look like the
+		// call to prepareExistingTableDescForIngestion. Currently, FK references
+		// misbehave when I tried to write the desc one at a time.
+		if len(newTableDescs) != 0 {
+			res, err := prepareNewTableDescsForIngestion(ctx, txn, p, newTableDescs, importDetails.ParentID)
+			if err != nil {
+				return err
+			}
+			for _, v := range res {
+				importDetails.Tables[newTableDescToIdx[v]].Desc = v
+			}
+		}
+
+		importDetails.PrepareComplete = true
+
+		// If we do not have pending schema changes on existing descriptors we can
+		// choose our Walltime (to IMPORT from) immediately. Otherwise, we have to
+		// wait for all nodes to see the same descriptor version before doing so.
+		if !hasExistingTables {
+			importDetails.Walltime = p.ExecCfg().Clock.Now().WallTime
+		} else {
+			importDetails.Walltime = 0
+		}
+
+		// Update the job once all descs have been prepared for ingestion.
+		if err = r.job.WithTxn(txn).SetDetails(ctx, importDetails); err != nil {
+			return err
```

Comment:
ugh, for some reason our linter doesn't like this (err check here then nil return below, and just wants you to return it)